pkg/agent/controllers/agent/agent.go

```diff
@@ -0,0 +1,204 @@
+
+	// Start the informer factories to begin populating the informer caches
+	klog.Info("Starting Agent controller")
+
+	// Wait for the caches to be synced before starting workers
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, a.servicesSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	go wait.Until(a.runServiceWorker, time.Second, stopCh)
+	wg.Wait()
+	klog.Info("Lighthouse agent workers started")
+	<-stopCh
+	klog.Info("Lighthouse Agent stopping")
+	return nil
+}
+
+func (a *Controller) enqueueService(obj interface{}) {
+	if key := a.getEnqueueKey(obj); key != "" {
+		klog.V(4).Infof("Enqueueing service %v for agent", key)
+		a.serviceWorkqueue.AddRateLimited(key)
+	}
+}
+
+func (a *Controller) getEnqueueKey(obj interface{}) string {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return ""
+	}
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return ""
+	}
+	if a.excludeNamespaces[ns] || excludeSvcList[name] {
+		return ""
+	}
+	return key
+}
+func (a *Controller) runServiceWorker() {
+	for a.processNextService() {
+
+	}
+}
+
+func (a *Controller) processNextService() bool {
+	obj, shutdown := a.serviceWorkqueue.Get()
+	if shutdown {
+		return false
+	}
+	err := func() error {
+		defer a.serviceWorkqueue.Done(obj)
+
+		key := obj.(string)
+		ns, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			a.serviceWorkqueue.Forget(obj)
+			return fmt.Errorf("error while splitting meta namespace key %s: %v", key, err)
+		}
+
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// Retrieve the latest version of object before attempting update
+			// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+			service, err := a.kubeClientSet.CoreV1().Services(ns).Get(name, metav1.GetOptions{})
+
+			if err != nil {
+				if errors.IsNotFound(err) {
+					// already deleted Forget and return
+					a.serviceWorkqueue.Forget(obj)
+				} else {
+					// could be API error, so we want to requeue
+					logAndRequeue(key, a.serviceWorkqueue)
+				}
+				return fmt.Errorf("error retrieving service %s: %v", name, err)
+			}
+
+			return a.serviceUpdater(service, key)
+		})
+		if retryErr != nil {
+			return retryErr
+		}
+
+		a.serviceWorkqueue.Forget(obj)
+		return nil
+	}()
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+func (a *Controller) handleUpdateService(old interface{}, newObj interface{}) {
+	// ServiceIp can't change without deleting service, so we only check for globalIp change
+	if key := a.getEnqueueKey(newObj); key != "" {
+		oldGlobalIp := getGlobalIpFromService(old.(*corev1.Service))
+		newGlobalIp := getGlobalIpFromService(newObj.(*corev1.Service))
+		if oldGlobalIp != newGlobalIp {
+			a.serviceWorkqueue.AddRateLimited(key)
+		}
```

Comment:
Update will also be called at each resync period, in which case the objects will likely be equal. With this check, auto resync is rendered essentially useless and we should just disable it, which is fine as long as we re-queue/retry on all failures.

Reply:
Good idea. This sounds preferable than auto-resync.